### PR TITLE
metal: more PP speedup for Falcon

### PR DESCRIPTION
**Pull request by ikawrakow**
TL;DR: For PP on metal, a significant fraction of the time is spent in the `K * Q` matrix multiplication. The PR is an attempt to speed up this matrix multiplication. The result is mostly useful for the Falcon model where we see a ~12% increase in performance.

If I turn the `f16 x f32` matrix multiplication operation into a no-op, I see a 51% increase in PP performance for Falcon-7B and 17% gain for LLaMA-7B on metal. This explains most of the PP performance gap between between Falcon and LLaMA. Most of the time spent in `f16 x f32` matrix multiplications is due to the `K * Q` matrix multiplication. On metal, the kernel utilized for most other matrix multiplications cannot be used because the `K` and `Q` tensors are not contiguous. Hence, a poor-men's matrix multiplication kernel is invoked, which becomes a bottleneck as the context/batch size increases. This PR adds a slightly better version, which I think is still very far from optimal. The effect is a ~12% speedup for Falcon-7B PP, along with some very minor gains for TG.

30-core M2 Max:

| model                          | backend    | test       |    /s (Master)   |      t/s (PR)    |  Speedup |
| ------------------------------ | ---------- | ---------- | ---------------: | ---------------: | -------: |
| Falcon 7B mostly F16           | Metal      | pp 512     |    380.18 ± 0.13 |    426.83 ± 0.26 |   1.123  |
| LLaMA 7B mostly F16            | Metal      | pp 512     |    539.46 ± 0.31 |    539.61 ± 0.37 |   1.000  |
| LLaMA 7B mostly Q8_0           | Metal      | pp 512     |    486.91 ± 0.18 |    486.83 ± 0.44 |   1.000  |
| LLaMA 7B mostly Q4_0           | Metal      | pp 512     |    495.91 ± 0.38 |    495.82 ± 0.17 |   1.000  |
| LLaMA 7B mostly Q4_1           | Metal      | pp 512     |    494.31 ± 0.26 |    494.52 ± 0.55 |   1.000  |
| LLaMA 7B mostly Q6_K           | Metal      | pp 512     |    413.82 ± 0.39 |    414.16 ± 0.23 |   1.001  |
| LLaMA 7B mostly Q5_K - Small   | Metal      | pp 512     |    413.35 ± 0.27 |    413.58 ± 0.12 |   1.000  |
| LLaMA 7B mostly Q4_K - Small   | Metal      | pp 512     |    438.64 ± 0.23 |    438.39 ± 0.18 |   1.000  |
| LLaMA 7B mostly Q3_K - Small   | Metal      | pp 512     |    417.91 ± 0.25 |    417.79 ± 0.25 |   1.000  |
| Falcon 7B mostly F16           | Metal      | tg 128     |     23.33 ± 0.04 |     23.51 ± 0.05 |   1.008  |
| LLaMA 7B mostly F16            | Metal      | tg 128     |     24.26 ± 0.04 |     24.24 ± 0.07 |   1.000  |
| LLaMA 7B mostly Q8_0           | Metal      | tg 128     |     40.16 ± 0.27 |     40.35 ± 0.25 |   1.005  |
| LLaMA 7B mostly Q4_0           | Metal      | tg 128     |     62.53 ± 0.06 |     63.34 ± 0.04 |   1.013  |
| LLaMA 7B mostly Q4_1           | Metal      | tg 128     |     58.88 ± 0.78 |     58.62 ± 0.02 |   0.996  |
| LLaMA 7B mostly Q6_K           | Metal      | tg 128     |     42.93 ± 0.48 |     43.21 ± 0.46 |   1.006  |
| LLaMA 7B mostly Q5_K - Small   | Metal      | tg 128     |     44.67 ± 0.03 |     45.79 ± 0.03 |   1.025  |
| LLaMA 7B mostly Q4_K - Small   | Metal      | tg 128     |     57.31 ± 0.44 |     57.86 ± 0.04 |   1.010  |
| LLaMA 7B mostly Q3_K - Small   | Metal      | tg 128     |     53.42 ± 0.24 |     53.77 ± 0.17 |   1.007  |

### CODE REVIEW

**Comment by ikawrakow** [comment]
Sure, copying `Q` to contiguous memory is an obvious approach, but I wanted to avoid the copy to not have the TG performance degradation. For LLaMA-7B the `Q` tensor is 128 x Nb x 32, so the amount of data to be copied becomes non-trivial when generating a long sequence of tokens.

**Comment by ggerganov** [comment]
During TG, `Nb == 1`, so the `Q` tensor is always the same size and it is small. Here is for example the graph for `n_past == 45` and `n_batch == 1`:

image

**Comment by ikawrakow** [comment]
Given #3168, I'm closing this PR.
